gix-worktree/src/index/status/worktree.rs

```diff
@@ -0,0 +1,272 @@
+            // TODO: check if any parent directory is a symlink
+            //       we need to use fs::Cache for that
+            Ok(metadata) if metadata.is_dir() => {
+                // index entries are normally only for files/symlinks
+                // if a file turned into a directory it was removed
+                // the only exception here are submodules which are
+                // part of the index despite being directories
+                //
+                // TODO: submodules:
+                //   if entry.mode.contains(Mode::COMMIT) &&
+                //     resolve_gitlink_ref(ce->name, "HEAD", &sub))
+                return Ok(Status::Removed);
+            }
+            Ok(metadata) => metadata,
+            Err(err) if err.kind() == io::ErrorKind::NotFound => return Ok(Status::Removed),
+            Err(err) => {
+                return Err(err.into());
+            }
+        };
+        if self.options.check_added && entry.flags.contains(index::entry::Flags::INTENT_TO_ADD) {
+            return Ok(Status::Added);
+        }
+        let new_stat = index::entry::Stat::from_fs(&metadata)?;
+        let executable_bit_changed =
+            match entry
+                .mode
+                .change_to_match_fs(&metadata, self.options.fs.symlink, self.options.fs.executable_bit)
+            {
+                Some(index::entry::mode::Change::Type { .. }) => return Ok(Status::TypeChange),
+                Some(index::entry::mode::Change::ExecutableBit) => true,
+                None => false,
+            };
+
+        // Here we implement racy-git. See racy-git.txt in the git documentation for a detailed documentation
+        // A file is racy if:
+        // * it's mtime is at/after the last index timestamp, it's entry stat information
+        //   matches the on-disk file but the file contents are actually modified
+        // * it's size is 0 (set after detecting a file was racy previously)
+        //
+        // The first case is detected below (by checking the timestamp if the file is marekd umodified).
+        // The second case is usually detected either because the on-disk file
+        // is not empty (hence the basic stat match fails) or by checking
+        // whether the size doesn't fit the oid (the oid of an empty blob is a constant)
+        let mut racy_clean = false;
+        if !executable_bit_changed
+            && new_stat.matches(&entry.stat, self.options.stat)
+            && (!entry.id.is_empty_blob() || entry.stat.size == 0)
+        {
+            racy_clean = new_stat.is_racy(self.timestamp, self.options.stat);
+            if !racy_clean {
+                return Ok(Status::Unchanged);
+            }
+        }
+
+        let file = WorktreeFile {
+            buf: &mut self.buf,
+            path: &worktree_path,
+            entry,
+            options: self.options,
+        };
+        let diff = diff.content_changed::<Error>(entry, metadata.len() as usize, file, |_| Ok(&[]))?;
```

Comment:
Yeah this is intentionally a placeholder right now. Probably should have left a comment :D The find callback will tie in with the fs cache (since that also needs to read from the ODB) and I wanted to wait for you there so I don't have to redo work if the API changes.

Reply:
Ah, right! Since access to the ODB is needed I have added the wiring at least, and I think it makes more sense than before and should be a decent basis for the `fs::Cache` integration.